sdks/python/apache_beam/io/iobase.py

```diff
@@ -992,6 +1012,43 @@ def expand(self, pcoll):
           'A sink must inherit iobase.Sink, iobase.NativeSink, '
           'or be a PTransform. Received : %r' % self.sink)
 
+  def _pubsub_write_payload(self):
+    return beam_runner_api_pb2.PubSubWritePayload(
+        topic=self.sink.full_topic,
+        id_attribute=self.sink.id_label,
+        timestamp_attribute=self.sink.timestamp_attribute)
+
+  def to_runner_api_parameter(self, context):
+    # type: (PipelineContext) -> Tuple[str, Any]
+    # Importing locally to prevent circular dependencies.
+    from apache_beam.io.gcp.pubsub import _PubSubSink
+    if isinstance(self.sink, _PubSubSink):
+      payload = self._pubsub_write_payload()
+      return (common_urns.composites.PUBSUB_WRITE.urn, payload)
+    else:
+      return super(Write, self).to_runner_api_parameter(context)
+
+  @staticmethod
+  @ptransform.PTransform.register_urn(
+      common_urns.composites.PUBSUB_WRITE.urn,
+      beam_runner_api_pb2.PubSubWritePayload)
+  def from_runner_api_parameter(ptransform, payload, unused_context):
+    # type: (Any, Any, PipelineContext) -> Write
+    if ptransform.spec.urn != common_urns.composites.PUBSUB_WRITE.urn:
+      raise ValueError(
+          'Write transform cannot be constructed for the given proto %r',
+          ptransform)
+
+    # Importing locally to prevent circular dependencies.
+    from apache_beam.io.gcp.pubsub import _PubSubSink
+    sink = _PubSubSink(
+        topic=payload.topic,
+        id_label=payload.id_attribute,
+        with_attributes=True,
```

Comment:
If `with_attributes` is always `True`, should we also remove `with_attributes` from `_PubSubSink` and  `_PubSubSource`? This change requires `_PubSubSource` only outputs `PubSubMessage` and `_PubSubSink` only accepts `PubSubMessage`.

Reply:
So, I thought that's the simplification you are working on, right ? I'm fine with these properties being removed form source/sink if they are always true.

This change simply updates protos (and corresponding encoding in Read/Write transforms) to match your updates to PubSub native source/sink.

Reply:
I only did partial cleanup to Write. We can have a followup PR to add more cleanup.
cc: @lukecwik

Reply:
Hmm, actually "with_attributes=True" is not really correct for current users since we do a runner API to and from round trip here: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/dataflow/dataflow_runner.py#L511

I think the next best course of action is to revert https://github.com/apache/beam/pull/12760 from the release and get this in once Boyuan's source/sink simplifications are in.

Reply:
I added 'with_attributes' property back. That should fix the failure here. We can later deprecate/remove it from the proto if we truly remove all it's usage including DirectRunner.

Reply:
Could you explain more about why it also affects DirecrRunner? I would prefer checking in cleanup for Sink/Source first then checking in clean version of proto. What do you think?

Reply:
I think the crux of the matter is that runners (both Dataflow and Direct) depend on pipeline->proto->pipleline transformation to preserve state. 
All runners depend on it here: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/pipeline.py#L514
Additional Dataflow performs a second round trip here: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/dataflow/dataflow_runner.py#L511

Additionally DirectRunner refers to 'with_attributes' property here which is expected to be preserved in such a transformation: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/direct/direct_runner.py#L406

My proposal it to remove other Dataflow specific state from the proto and keep `with_attributes` which is needed to preserve state of the transforms today.

Reply:
After an offline discussion, we'll get the revert (https://github.com/apache/beam/pull/12914) in the unblock the 2.25.0 release and get this PR in after Boyuan's upcoming refactoring efforts.